### ISSUE TITLE
Minimal Implementation

### ISSUE THREAD

**Issue by kroggen** [Haiyang-W/TokenFormer#5]
I tried to implement a minimal version of this code for learning purposes:

https://github.com/kroggen/tokenformer-minimal

But the output is gibberish... Something is not right.

If someone can figure out what is the issue, please share

**Comment by Haiyang-W** [comment]
> I tried to implement a minimal version of this code for learning purposes:
> 
> https://github.com/kroggen/tokenformer-minimal
> 
> But the output is gibberish... Something is not right.
> 
> If someone can figure out what is the issue, please share

Thanks for your interest! The training code is still being organized. I’ve been sick these past few days and haven’t had the chance to finish it, but the training code should be available next week I hope.

**Comment by Haiyang-W** [comment]
Actually, the code is quite simple. In a reliable LLM codebase, you just replace the transformer's linear projection with pattention. I think the performance should be fine, and pay attention to initialization and similar aspects. However, we’re only using the most basic settings.

**Comment by kroggen** [comment]
It is not "just replace", because the positional embeddings are applied to only 25% of the inputs if I am understanding the code correctly.

And it may have some other tricks made by megatron, deepseed and other modules you are using (and I am trying to get rid of)

The minimal code I made is for inference only, using your models, but it still does not work...

The goal is to port it to tinygrad later

**Comment by kroggen** [comment]
The paper says:

> τ is the scale factor, which is set to √n by default

But the code uses `1` as the scale factor for Pattention ([check here](https://github.com/Haiyang-W/TokenFormer/blob/main/megatron/model/tokenformer.py#L114))

Also the `attn_bias` is a no-op ([multiplying by 1](https://github.com/Haiyang-W/TokenFormer/blob/main/megatron/model/tokenformer.py#L127))

**Comment by Haiyang-W** [comment]
> It is not "just replace", because the positional embeddings are applied to only 25% of the inputs if I am understanding the code correctly.
This part is the trick used in Pythia, irrelevant to our paper. So you can think of it as a reliable transformer-based LLM framework where you just need to replace the linear projection? 

However, if you want to match the inference, you may need to align all the tricks used in the code. These are some tricks used by Pythia, megatron and deepseed. :)

If you have any questions, please feel free to leave a message, and I’ll do my best to respond promptly. :)

Best,
Haiyang

**Comment by kroggen** [comment]
OK.

Some notes:

The [norm_factor](https://github.com/Haiyang-W/TokenFormer/blob/main/megatron/model/tokenformer.py#L236-L242) appears to not be used anywhere

The [attn_bias](https://github.com/Haiyang-W/TokenFormer/blob/main/megatron/model/tokenformer.py#L116-L127) appears to be an attention mask

**Comment by kroggen** [comment]
It is working now! :slightly_smiling_face:

**Comment by Haiyang-W** [comment]
> It is working now! 🙂

That's great! 🙂

**Comment by Holmes-GU** [comment]
> It is working now! 🙂

Hi, 

After reading about your issues, I learned much about understanding the codes further. I have two questions and hope you can give some help:

1) Do you align the reported results or get similar results? 
2) And what tricks make the results better?

Thanks for your time and efforts.

**Comment by kroggen** [comment]
Hello @Holmes-GU 

Unhappily I was not able to run this main repo to compare the results. There is no easy way to run inference on arbitrary prompts here, only support to run the evals

Check #7
